src/QAPI.h

```diff
@@ -0,0 +1,151 @@
+/* Not any company's property but Public-Domain
+* Do with source-code as you will. No requirement to keep this
+* header if need to use it/change it/ or do whatever with it
+*
+* Note that there is No guarantee that this code will work
+* and I take no responsibility for this code and any problems you
+* might get if using it.
+*
+* Code & platform dependent issues with it was originally
+* published at http://www.kjellkod.cc/threadsafecircularBase
+* 2012-16-19  @author Kjell Hedström, hedstrom@kjellkod.cc
+*
+* Modified and inspired from KjellKod's code at:
+* https://github.com/KjellKod/Q/src/q/q_api.hpp
+*/
+
+#pragma once
+
+#include <tuple>
+#include <memory>
+#include <TimeStats.h>
+#include <TriggerTimeStats.h>
+#include <q/spsc.hpp>
+#include <q/mpmc.hpp>
+
+namespace QAPI {
+   // Base Queue API without pop() and push()
+   // This follows the 'tail' first design on FIFO
+   // http://en.wikipedia.org/wiki/FIFO#Head_or_tail_first
+   // This implementation follows "pop on head", "push on tail"
+   template<typename QType>
+   struct Base {
+      Base(std::shared_ptr<QType> q)
+         : mQueueStorage(q)
+         , mQueueRef(*(q.get())) {
+      }
+      bool empty() const { return mQueueRef.empty();}
+      bool full() const { return mQueueRef.full(); }
+      size_t capacity() const { return mQueueRef.capacity(); }
+      size_t capacity_free() const { return mQueueRef.capacity_free(); }
+      size_t size() const { return mQueueRef.size(); }
+      // GCC BUG:  std::atomic<size_t>{}.is_lock_free(); is not implemented in gcc 4.8.5
+      // It is implemented in later versions (gc 5.3 and newer)
+      //bool lock_free() const { return mQueueRef.lock_free(); } 
+      size_t usage() const { return mQueueRef.usage(); }
+
+      std::shared_ptr<QType> mQueueStorage;
+      QType& mQueueRef;
+   };
+
+
+   // push() + base Queue API
+   template<typename QType>
+   struct Sender : public Base<QType> {
+    public:
+      Sender(std::shared_ptr<QType> q): Base<QType>(q) {}
+      virtual ~Sender() = default;
+
+      template<typename Element>
+      bool push(Element& item) {
+         TriggerTimeStats trigger(mStats);
+         auto result = Base<QType>::mQueueRef.push(item);
+         if (!result) {
+            trigger.Skip();
+         }
+         return result;
+      }
+      TimeStats mStats;
+   };
+
+
+   namespace sfinae {
+      // SFINAE: Substitution Failure Is Not An Error
+      // Decide at compile time what function signature to use
+      // 1. If 'wait_and_pop' exists in the queue it uses that
+      // 2. If only 'pop' exists it implements 'wait_and_pop' expected
+      // -- FYI: The wait is set to increments of 100 ns
+      template <typename T, typename Element>
+      bool wrapper(T& t, Element& e, std::chrono::milliseconds max_wait) {
+         using milliseconds = std::chrono::milliseconds;
+         using clock = std::chrono::steady_clock;
+         using namespace std::chrono_literals;
```

Comment:
Good question. 

It should be kept within the function in my opinion with due to that  it is only used within the function.  I.e. similar to our practice of declaring variables close to its usage. 
If it would be used all over then I agree it should be within the namespace. 

Another benefit of this is that it doesn't pollute the `sfinae` namespace which would happen if you put the `using` statements within the namespace instead of the function.